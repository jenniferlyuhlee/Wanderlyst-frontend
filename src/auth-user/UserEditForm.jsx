import { useState, useContext } from "react";
import { NavLink } from "react-router-dom";
import WanderlystApi from "../utils/api";
import cleanData from "../helpers/cleanData";
import UserContext from "./UserContext";
import Alert from "../shared/Alert";

/** User Edit Form component
 * Allow users to edit basic information or delete account
 */

function UserEditForm(){
    const { currUser, setCurrUser, logout} = useContext(UserContext);
    
    // states: formData, formErrors, saveConfirmed
    const [formData, setFormData] = useState({
        username: currUser.username,
        password: "",
        firstName: currUser.firstName,
        lastName: currUser.lastName,
        location: currUser.location || "",
        bio: currUser.bio || "",
        profilePic: currUser.profilePic || ""
    })
    const [formErrors, setFormErrors] = useState(null);
    const [saveConfirmed, setSaveConfirmed] = useState(false);
    const [settingErrors, setSettingErrors] = useState(null);

    // general callback to update targeted form data
    function handleChange (e) {
        const { name, value } = e.target;
        // default formData to empty string if value is undefined from currUser
        setFormData(data => ({...data, [name]: value || ""}));
    }

    // form submission handler
    async function handleSubmit(e){
        e.preventDefault();
        let updatedUser;
        try{
            // removes duplicate confirmation password from data
            delete formData.confirmPassword  
            const cleanedFormData = cleanData(formData);
            updatedUser = await WanderlystApi.updateUser(currUser.username, cleanedFormData);

        }
        catch(err){
            setFormErrors(err)
            return;
        }
        // resets profile form to display updated info
        setFormErrors(null);
        setSaveConfirmed(true);
        setCurrUser({
            username: updatedUser.username, 
            firstName: updatedUser.firstName,
            lastName: updatedUser.lastName,
            location: updatedUser.location,
            bio: updatedUser.bio,
            profilePic: updatedUser.profilePic,
            isAdmin: updatedUser.isAdmin, 
            likes: currUser.likes});
    }
    

       // delete user handler
    async function handleDelete(){
        try{
            alert(`Deleted account!`)
            const deleted = await WanderlystApi.deleteUser(currUser.username)
            if (deleted){
                logout();
            }
        }
        catch(err){
            setSettingErrors(err);
        }
    }

    return (
        <div className = "container">
            <NavLink to ={`/users/profile/${currUser.username}`}
            className = "btn rounded-pill btn-secondary mt-4 mb-3">
                Return
            </NavLink>
        <div className = "card px-5 mb-5">
            <h2>Edit User Info</h2>
            <form onSubmit = {handleSubmit} className = "row m-1">
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "firstName"
                        className = "form-control"
                        name = "firstName"
                        placeholder = "firstName"
                        value = {formData.firstName}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "firstName">First Name</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "lastName"
                        className = "form-control"
                        name = "lastName"
                        placeholder = "lastName"
                        value = {formData.lastName}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "lastName">Last Name</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "location"
                        className = "form-control"
                        name = "location"
                        placeholder = "location"
                        value = {formData.location}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "location">Location</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "bio"
                        className = "form-control"
                        name = "bio"
                        placeholder = "bio"
                        value = {formData.bio}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "bio">Bio</label>
                </div>
                <div className = "form-floating col-12 mb-2 p-0">
                    <input id = "profilePic"
                        className = "form-control"
                        name = "profilePic"
                        placeholder = "profilePic"
                        value = {formData.profilePic}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "profilePic">Profile Pic URL</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "username"
                        className = "form-control"
                        name = "username"
                        placeholder = "username"
                        value = {formData.username}
                        onChange = {handleChange}
                        disabled
                    />
                    <label htmlFor = "username">Username</label>
                </div>
                <div className = "form-floating col-6 mb-2 p-0">
                    <input id = "password"
                        className = "form-control"
                        name = "password"
                        type = "password"
                        placeholder = "password"
                        value = {formData.password}
                        onChange = {handleChange}
                    />
                    <label htmlFor = "password">New Password</label>
                </div>
                {formErrors?
                    <Alert type = "danger" messages = {formErrors}/>
                    :
                    null
                }
                <button className = "btn rounded-pill btn-lg btn-primary">
                    Save
                </button>
                {saveConfirmed?
                    <Alert type="success" messages={["Updated successfully."]} />
                    :
                    null
                }
            </form>
            <hr  className = "my-3"/>
            <div>
                <h3>Settings</h3>
                <button onClick={handleDelete} className="btn rounded-pill btn-lg btn-danger">
                    Delete Account
                </button>
                {settingErrors? 
                    <Alert type="warning" messages={["Unable to proceed with your request. Try again later."]} />
                    :
                    null
                }
            </div>
        

        </div>
        </div>
    )
}
export default UserEditForm;