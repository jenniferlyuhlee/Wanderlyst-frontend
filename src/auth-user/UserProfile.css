.Profile-title{
    text-align: center;
}
.Profile-header{
    padding: 0 3rem;
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.Profile-header-group{
    display: flex;
    align-items: center;
}
.Profile-img{
    height: 18vw;
    width: 18vw;
    border-radius: 50%;
}
.Profile-info{
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 2rem 3rem;
}

@media (max-width: 800px){
    .Profile-header{
        flex-direction: column;
        padding: 0;
    }
    .Profile-header-group{
        flex-direction: column;
    }
    .Profile-img{
        height: 150px;
        width: 150px;
    }
    .Profile-info{
        margin-top: 1rem;
        padding: 0;
    }
}

.Profile-separator{
    margin: 2rem 0 0 0;
}

@media (max-width: 1200px){
    .Profile-separator{
        margin: 0;
    }
}

.Profile-add-itin{
    color: rgba(102, 209, 149);
}

.Profile-itins-menu{
    display: flex;
    justify-content: space-around;
}

.itin-menu{
    border-radius: 0;
    padding: 0.5rem 0;
    width: 100%;
    background-color: white;
    border: 0;
    border-bottom: solid white 3px;
    color: #888;
    transition: color, 0.2s ease-out;
}
.itin-menu:hover{
    color: #555555;
}

.active-tab{
    pointer-events: none;
    cursor: not-allowed;
    color: rgba(102, 209, 149);
    border-bottom: 3px solid;
}

.bi{
    font-size: 3rem;
}

.bi-geo-alt-fill{
    font-size: 1rem;
}