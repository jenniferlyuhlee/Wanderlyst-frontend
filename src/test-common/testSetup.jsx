import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };