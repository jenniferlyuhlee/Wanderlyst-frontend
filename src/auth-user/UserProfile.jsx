import React, { useState, useEffect, useContext} from "react";
import { NavLink, useParams } from "react-router-dom";
import WanderlystApi from "../utils/api";
import UserContext from "./UserContext";
import dateConvert from "../helpers/dateConvert";
import ItinCard from "../itineraries/ItinCard";
import Loading from "../shared/Loading";
import "./UserProfile.css"

/** User Profile Page Component
 * Displays user details and associated itineraries (created/likes)
 */

function UserProfile(){
    const { currUser, likes} = useContext(UserContext);
    const { username } = useParams();

    // state set to null to use loading spinner
    const [user, setUser] = useState(null);
    // error state for invalid param user username
    const [error, setError] = useState(false);
    // toggle display of itineraries/likes
    const [showItins, setShowItins] = useState(true);

    useEffect(function getUserDetails() {
        async function getUser(){
            try{
                setUser(await WanderlystApi.getUser(username));
            }
            catch(err){
                console.error("User not found");
                setError(true);
            }
        }
        getUser();
    }, [username, likes]);

    // toggle itineraries / likes
    function toggleDisplay(){
        setShowItins(state => !state)
    }

    if(error) return <p className="NotFound"><i>Sorry, this user could not be found.</i></p>

    if(!user) return <Loading />

    return(
        <div className = "container pt-2 mb-5 body-cont">
            <h1 className = "Profile-title">@{user.username}</h1>
            <div className = "Profile-header">
                <div className = "Profile-header-group">
                    <img className = "Profile-img" src = 
                    {user.profilePic ? `${user.profilePic}`
                        : "/src/assets/default_profile.jpg"
                    }
                    />
                    <div className = "Profile-info">
                        <h4 className = "Profile-name">{user.firstName} {user.lastName}</h4>
                        {user.location ? 
                            <p>
                                <i className="bi bi-geo-alt-fill"></i> {user.location}
                            </p> 
                        : null}
                        {user.bio ? <p>{user.bio}</p> : null}
                        <small className = "mt-2"><i>Joined {dateConvert(user.createdAt)}</i></small>
                        {currUser.username === user.username ? 
                        <NavLink to={`/users/profile/edit`}
                            className = "mt-3 btn btn-sm rounded-pill btn-secondary">
                            Edit Profile
                        </NavLink>
                        : null}
                    </div>
                </div>
                {currUser.username === user.username ? 
                    <NavLink to="/itineraries/new" className="Profile-add-itin">
                        <i className="bi bi-plus-circle-fill"></i>
                    </NavLink>
                : null}
            </div>
            <hr className = "Profile-separator"/>
            <div>
                <div className = "Profile-itins-menu">
                    <button onClick={toggleDisplay}
                    className = {`itin-menu ${showItins? "active-tab" : ""}`}>
                        <i className = {`bi bi-map${showItins? "-fill" : ""}`}></i>
                    </button>
                    <button onClick={toggleDisplay}
                    className = {`itin-menu ${showItins? "" : "active-tab"}`}>
                        <i className = {`bi bi-bookmark-heart${showItins? "" : "-fill"}`}></i>
                    </button>
                </div>
                {showItins ?
                <div className = "User-itins m-4">
                    <div className = "mt-2">
                    {user.itineraries.length > 0 ?
                         user.itineraries.map(i => (
                            <ItinCard key = {i.id} itinerary={i}/>
                        ))
                    :
                        <div className = "text-center">
                            <i>
                                {currUser.username === user.username? "Start sharing itineraries!" : `${user.username} has no itineraries yet.`}
                            </i>
                        </div>
                    }
                    </div>
                </div>
                :
                <div className = "userLikes m-4">                
                    <div className = "mt-2">
                    {user.likes.length > 0 ?
                        user.likes.map(i => (
                            <ItinCard key = {i.id} itinerary={i}/>
                        ))
                    :
                    <div className = "text-center">
                        <i>
                            {currUser.username === user.username? "Start exploring and liking itineraries!" : `${user.username} has no likes yet.`}
                        </i>
                    </div>
                    }
                    </div> 
                </div>
                }
            </div>
        </div>
    )
}

export default UserProfile;