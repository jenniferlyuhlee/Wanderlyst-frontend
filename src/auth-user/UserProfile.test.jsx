import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import WanderlystApi from "../utils/api";
import UserProfile from "./UserProfile";
import UserEditForm from "./UserEditForm";
import { beforeEach, afterEach, expect, describe, test } from "vitest";

describe("UserProfile Component", () => {
    // mock api method to return testadmin info
    beforeEach(() => {
        vi.spyOn(WanderlystApi, "getUser").mockResolvedValue({
            username: "testadmin",
            firstName: "testfirst",
            lastName: "testlast",
            location: "testlocation",
            bio: "testbio",
            profilePic: null,
            isAdmin: true,
            createdAt: '2024-10-01',
            likes: [],
            itineraries: []
        });
    });
    
    // Restore mocks after tests
    afterEach(() => {
        vi.restoreAllMocks();
    });

    test("renders without crashing", async () => {
        await waitFor(() => {render(
            <MemoryRouter initialEntries={['/users/profile/testadmin']}>
                <UserProvider>
                    <UserProfile />
                </UserProvider>
            </MemoryRouter>)
        });
    });
            
    test("matches snapshot", async () => {
        const {asFragment, findByText} = render(
            <MemoryRouter initialEntries={['/users/profile/testadmin']}>
                <UserProvider>
                    <UserProfile />
                </UserProvider>
            </MemoryRouter>)

        // wait for element that shows user data has loaded
        await findByText('testfirst testlast');
        expect(asFragment()).toMatchSnapshot()
    });

    test("works as expected: toggling displays", async () => {
        const { container, getByText } = render(
            <MemoryRouter initialEntries={['/users/profile/testadmin']}>
                <UserProvider>
                    <UserProfile />
                </UserProvider>
            </MemoryRouter>)
        
   
        // user itineraries tab shows on initial load
        await waitFor(() => {
            expect(getByText('Start sharing itineraries!')).toBeInTheDocument()
        });
     
        // click likes tab
        const likesBtn = container.querySelector(`.bi-bookmark-heart`).parentElement;
        fireEvent.click(likesBtn);

        // user likes tab now displays
        await waitFor(() => {
            expect(getByText('Start exploring and liking itineraries!')).toBeInTheDocument()
        });
    });

    test("edit profile button navigates to edit form", async () => {
        const {getByText} = render(
            <MemoryRouter initialEntries={['/users/profile/testadmin']}>
                <UserProvider>
                    <Routes>
                        <Route path="/users/profile/testadmin" element={<UserProfile />} />
                        <Route path="/users/profile/edit" element={<div>User Edit Form</div>}/>
                    </Routes>
                </UserProvider>
            </MemoryRouter>)
        
        await waitFor(() => {
            expect(getByText('@testadmin')).toBeInTheDocument()
        });
        
        // click edit profile button
        fireEvent.click(getByText("Edit Profile"));

        // check for reroute
        expect(getByText("User Edit Form")).toBeInTheDocument();
    })
});