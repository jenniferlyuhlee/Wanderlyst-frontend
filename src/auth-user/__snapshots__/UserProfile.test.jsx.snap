// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserProfile Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container pt-2 mb-5 body-cont"
  >
    <h1
      class="Profile-title"
    >
      @testadmin
    </h1>
    <div
      class="Profile-header"
    >
      <div
        class="Profile-header-group"
      >
        <img
          class="Profile-img"
          src="/src/assets/default_profile.jpg"
        />
        <div
          class="Profile-info"
        >
          <h4
            class="Profile-name"
          >
            testfirst testlast
          </h4>
          <p>
            <i
              class="bi bi-geo-alt-fill"
            />
             testlocation
          </p>
          <p>
            testbio
          </p>
          <small
            class="mt-2"
          >
            <i>
              Joined Sep 2024
            </i>
          </small>
          <a
            class="mt-3 btn btn-sm rounded-pill btn-secondary"
            href="/users/profile/edit"
          >
            Edit Profile
          </a>
        </div>
      </div>
      <a
        class="Profile-add-itin"
        href="/itineraries/new"
      >
        <i
          class="bi bi-plus-circle-fill"
        />
      </a>
    </div>
    <hr
      class="Profile-separator"
    />
    <div>
      <div
        class="Profile-itins-menu"
      >
        <button
          class="itin-menu active-tab"
        >
          <i
            class="bi bi-map-fill"
          />
        </button>
        <button
          class="itin-menu "
        >
          <i
            class="bi bi-bookmark-heart"
          />
        </button>
      </div>
      <div
        class="User-itins m-4"
      >
        <div
          class="mt-2"
        >
          <div
            class="text-center"
          >
            <i>
              Start sharing itineraries!
            </i>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
